File: apps/demo-be-kotlin-ktor/src/test/kotlin/com/organiclever/demoktkt/unit/steps/UnitTokenManagementSteps.kt
package com.organiclever.demoktkt.unit.steps

import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertTrue

class UnitTokenManagementSteps {

  @When("alice decodes her access token payload")
  fun aliceDecodesHerAccessTokenPayload() {
    val token = UnitTestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val (status, body) = UnitHttpHelper.get("/api/v1/tokens/claims", token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @Then("the token should contain a non-null {string} claim")
  fun theTokenShouldContainNonNullClaim(claim: String) {
    UnitJsonHelper.assertNonNull(UnitTestWorld.lastResponseBody, claim)
  }

  @When("^the client sends GET /\\.well-known/jwks\\.json$")
  fun theClientSendsGetJwks() {
    val (status, body) = UnitHttpHelper.get("/.well-known/jwks.json")
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @Then("^the response body should contain at least one key in the \"keys\" array$")
  fun theResponseBodyShouldContainAtLeastOneKeyInKeysArray() {
    val body = UnitTestWorld.lastResponseBody
    assertTrue(body.contains("keys"), "Expected 'keys' in response: $body")
    assertTrue(body.contains("kty"), "Expected at least one key object in: $body")
  }

  @Then("alice's access token should be recorded as revoked")
  fun alicesAccessTokenShouldBeRecordedAsRevoked() {
    val token = UnitTestWorld.accessTokens["alice"] ?: error("alice has no access token")
    val decoded = UnitTestWorld.jwtService.decodeTokenUnchecked(token)
    val jti = decoded?.getClaim("jti")?.asString() ?: error("no jti in token")
    val isRevoked = runBlocking { UnitTestWorld.tokenRepo.isRevoked(jti) }
    assertTrue(isRevoked, "Expected alice's token jti=$jti to be revoked")
  }
}

File: apps/demo-be-csharp-aspnetcore/tests/DemoBeCsas.Tests/Unit/Steps/UnitTokenManagementSteps.cs
using System.IdentityModel.Tokens.Jwt;
using System.Text.Json;
using FluentAssertions;
using Reqnroll;
using Xunit;

namespace DemoBeCsas.Tests.Unit.Steps;

/// <summary>
/// Unit-level step bindings for token management and JWKS scenarios.
/// Mirrors <c>Integration.Steps.TokenManagementSteps</c> but scoped to scenarios tagged
/// <c>@unit</c> and uses <see cref="UnitServiceLayer"/> with in-memory repositories.
/// </summary>
[Binding]
[Trait("Category", "Unit")]
[Scope(Tag = "unit")]
public class UnitTokenManagementSteps(UnitServiceLayer svc, UnitSharedState state)
{
    // ─────────────────────────────────────────────────────────────
    // When steps
    // ─────────────────────────────────────────────────────────────

    [When(@"^alice decodes her access token payload$")]
    public void WhenAliceDecodesToken()
    {
        state.AccessToken.Should().NotBeNull("access token should be stored");
    }

    [When(@"^the client sends GET /\.well-known/jwks\.json$")]
    public void WhenGetJwks()
    {
        state.LastResponse = svc.GetJwks();
    }

    // ─────────────────────────────────────────────────────────────
    // Then steps
    // ─────────────────────────────────────────────────────────────

    [Then(@"^the token should contain a non-null ""([^""]+)"" claim$")]
    public void ThenTokenContainsClaim(string claimName)
    {
        state.AccessToken.Should().NotBeNull("access token should be stored");
        var handler = new JwtSecurityTokenHandler();
        var jwt = handler.ReadJwtToken(state.AccessToken!);

        if (claimName == "sub")
        {
            jwt.Subject.Should().NotBeNullOrEmpty($"Token should contain '{claimName}' claim");
        }
        else if (claimName == "iss")
        {
            jwt.Claims.Should().NotBeEmpty("Token should have claims");
        }
        else
        {
            var claimType = claimName switch
            {
                "jti" => JwtRegisteredClaimNames.Jti,
                "iat" => JwtRegisteredClaimNames.Iat,
                _ => claimName,
            };
            var claim = jwt.Claims.FirstOrDefault(c => c.Type == claimType || c.Type == claimName);
            claim.Should().NotBeNull($"Token should contain '{claimName}' claim");
        }
    }

    [Then(@"^the response body should contain at least one key in the ""keys"" array$")]
    public void ThenJwksContainsKeys()
    {
        state.LastResponse.Should().NotBeNull();
        var body = state.LastResponse!.Body;
        using var doc = JsonDocument.Parse(body);
        doc.RootElement.TryGetProperty("keys", out var keys).Should().BeTrue($"'keys' not found in: {body}");
        keys.GetArrayLength().Should().BeGreaterThan(0, "JWKS should contain at least one key");
    }
}
